### DISCUSSION TITLE
Loading historical data from api to bars

### DISCUSSION THREAD

**Discussion by OtabekEshpulatov** [tradingview/lightweight-charts#1341] [General]
I'm currently building my platform that consumes data from database and draws a chart to the screen using lightweightcharts library. Adding a new bar to candleseries is not a problem, but whenever user scrolls to the right, wanting to expose more data, I need to load data from database and insert it into candleseries. However, in the documentation, it says inserting a new bar that is older than the last bar in the series is not possible. But I want my users to be able to see historical data while loading it dynamically from my api. How can I solve this problem? It would be very helpful if you can give any advice. 

( series.setData() is not my option)

**Comment by SlicedSilver** [comment]
Could you explain why using `setData` wouldn't be a viable option? 

Here is a quick example on how you could use setData to update the entire series with the extra data: https://codesandbox.io/p/sandbox/lightweight-charts-ts-load-earlier-data-l57lo2

Related answers:
- https://stackoverflow.com/questions/75132897/fetching-backward-data-make-the-chart-stuck-for-a-while-with-lightweight-chart/75133235#75133235
- https://github.com/tradingview/lightweight-charts/issues/1340

**Reply by OtabekEshpulatov** [reply]
because i want to stream real time data from api. I think, they might interfere with one another

**Reply by OtabekEshpulatov** [reply]
what advice would you give me in this situation when i want to stream real time data and enable users to see the history by loading bars dynamically?

**Reply by SlicedSilver** [reply]
I've updated the [linked example](https://codesandbox.io/p/sandbox/lightweight-charts-ts-load-earlier-data-l57lo2) to include both loading historical data, and simulating realtime updates.

Essentially you could keep your own copy of the current chart data which you update at the same time as you update the chart (using the `update()` method) so that if you need to load earlier data then you can just combine the older data with your copy of the current data and then apply that to the chart using `setData()`.

So you only use `setData` if you need to update old data but you ensure that when you do use setData that it has all the current data.
